test/scale/scale_tester.go

```diff
@@ -0,0 +1,292 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scale
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	scale "k8s.io/api/autoscaling/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+	testtypes "sigs.k8s.io/azuredisk-csi-driver/test/types"
+	nodeutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/node"
+	podutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/pod"
+	volutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/volume"
+)
+
+type PodSchedulingWithPVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+func (t *PodSchedulingWithPVScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	totalNumberOfPods := t.Replicas
+	var oneMinPodsRunning,
+		twoMinPodsRunning,
+		threeMinPodsRunning,
+		fiveMinPodsRunning,
+		tenMinPodsRunning = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	var oneMinPodsDeleted,
+		twoMinPodsDeleted,
+		threeMinPodsDeleted,
+		fiveMinPodsDeleted,
+		tenMinPodsDeleted = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+
+	startScaleOut := time.Now()
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+	statefulSet, cleanupStatefulSet := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, t.Replicas, t.StorageClassParameters, &tStorageClass)
+	for i := range cleanupStatefulSet {
+		i := i
+		defer cleanupStatefulSet[i]()
+	}
+	statefulSet.CreateWithoutWaiting()
+	numberOfRunningPods := 0
+	ticker := time.NewTicker(time.Minute)
+	tickerStop := make(chan bool)
+	go func() {
+		time.Sleep(11 * time.Minute)
+		tickerStop <- true
+	}()
+	tickerCount := 0
+	stopLoop := false
+	for {
+		select {
+		case <-tickerStop:
+			stopLoop = true
+
+		case <-ticker.C:
+			tickerCount = tickerCount + 1
```

Comment:
ticker.C is going to report the time after every tick not every minute. We can't use it to count minutes like this.

Reply:
In this case 1 tick == 1 minute.

Reply:
Didn't realize you have passed Min to NewTicker for the interval.

Reply:
Using the tickerCount as a counter to measure the minutes since the ticker will tick after a minute interval. Not using ticker.C for reporting minutes.